File: FSUIKit/Classes/FSImage.m
//
//  FSImage.m
//  FSCalculator
//
//  Created by fudongdong on 2018/12/20.
//

#import "FSImage.h"
#import "FSCalculator.h"

@implementation FSImage

+ (UIImage *)imageFromColor:(UIColor *)color {
    return [self imageFromColor:color size:CGSizeMake(1, 10)];
}

+ (UIImage *)imageFromColor:(UIColor *)color size:(CGSize)size {
    if (![color isKindOfClass:UIColor.class]) {
        return nil;
    }
    CGRect rect = CGRectMake(0, 0, size.width, size.height);
    UIGraphicsBeginImageContext(rect.size);
    CGContextRef context = UIGraphicsGetCurrentContext();
    CGContextSetFillColorWithColor(context, [color CGColor]);
    CGContextFillRect(context, rect);
    UIImage *img = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    return img;
}

// 高德导航蚯蚓线
+ (UIImage *)imageWithSize1:(CGSize)size
                 direction:(BOOL)isHorizon
                    ratios:(NSArray<NSNumber *> *)ratios
                    colors:(UIColor *)colors,...NS_REQUIRES_NIL_TERMINATION {
    if (!([ratios isKindOfClass:NSArray.class] && ratios.count)) {
        return nil;
    }
    
    CGRect rect = CGRectMake(0, 0, size.width, size.height);
    UIGraphicsBeginImageContext(rect.size);
    CGContextRef context = UIGraphicsGetCurrentContext();
    
    Class _class_NSNumber = NSNumber.class;
    Class _class_Color = UIColor.class;
    CGFloat sum = 0;
    
    va_list args;
    va_start(args, colors);
    for (int x = 0; x < ratios.count; x ++) {
        NSNumber *ratio = ratios[x];
        if (![ratio isKindOfClass:_class_NSNumber]) {
            return nil;
        }
        UIColor *color = va_arg(args, UIColor *);
        if (![color isKindOfClass:_class_Color]) {
            return nil;
        }
        CGFloat fratio = ratio.floatValue;
        
        CGContextSetFillColorWithColor(context, color.CGColor);
        if (isHorizon) {
            CGContextFillRect(context, CGRectMake(0, sum * size.height, size.width, fratio * size.height));
        }else{
            CGContextFillRect(context, CGRectMake(sum * size.width, 0, fratio * size.width, size.height));
        }
        
        sum += fratio;
    }
    va_end(args);
    
    UIImage *img = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    return img;
}

// 高德导航蚯蚓线
+ (UIImage *)imageWithSize:(CGSize)size
                 direction:(BOOL)isHorizon
                offsetLeft:(CGFloat)left
               offsetRight:(CGFloat)right
                    ratios:(NSArray<NSNumber *> *)ratios
                    colors:(NSArray<UIColor *> *)colors {
    if (!([ratios isKindOfClass:NSArray.class] && ratios.count)) {
        return nil;
    }
    if (!([ratios isKindOfClass:NSArray.class] && ratios.count)) {
        return nil;
    }
    if (ratios.count != colors.count) {
        return nil;
    }
    
    CGRect rect = CGRectMake(0, 0, size.width, size.height);
    UIGraphicsBeginImageContext(rect.size);
    CGContextRef context = UIGraphicsGetCurrentContext();
    
    Class _class_NSNumber = NSNumber.class;
    Class _class_Color = UIColor.class;
    CGFloat sum = 0;
    
    for (int x = 0; x < ratios.count; x ++) {
        NSNumber *ratio = ratios[x];
        if (![ratio isKindOfClass:_class_NSNumber]) {
            return nil;
        }
        UIColor *color = colors[x];
        if (![color isKindOfClass:_class_Color]) {
            return nil;
        }
        CGFloat fratio = ratio.floatValue;
        
        CGContextSetFillColorWithColor(context, color.CGColor);
        if (isHorizon) {
            CGContextFillRect(context, CGRectMake(0, sum * size.height, size.width, fratio * size.height));
        }else{
            CGContextFillRect(context, CGRectMake(sum * size.width, 0, fratio * size.width, size.height));
        }
        
        sum += fratio;
    }
    
    UIImage *img = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    return img;
}

+ (UIImage *)imageWithSize:(CGSize)size
           backgroundColor:(UIColor *)backgroundColor
                 mainColor:(UIColor *)mainColor
               marginColor:(UIColor *)marginColor {
    CGRect rect = CGRectMake(0, 0, size.width, size.height);
    Class _class_Color = UIColor.class;

    UIGraphicsBeginImageContext(rect.size);
    CGContextRef context = UIGraphicsGetCurrentContext();
    
    if ([backgroundColor isKindOfClass:_class_Color]) {
        CGContextSetFillColorWithColor(context, backgroundColor.CGColor);
        CGContextFillRect(context, CGRectMake(0, 0, size.width * 0.25,size.height));
        CGContextFillRect(context, CGRectMake(size.width * 0.75, 0, size.width * 0.25, size.height));
    }
    
    if ([mainColor isKindOfClass:_class_Color]) {
        CGContextSetFillColorWithColor(context, mainColor.CGColor);
        CGContextFillRect(context, CGRectMake(size.width * 0.3125, 0, size.width * 0.375,size.height));
    }
    
    if ([marginColor isKindOfClass:_class_Color]) {
        CGContextSetFillColorWithColor(context, marginColor.CGColor);
        CGContextFillRect(context, CGRectMake(size.width * 0.25, 0, size.width * 0.0625,size.height));
        CGContextFillRect(context, CGRectMake(size.width * 0.6875, 0, size.width * 0.0625, size.height));
    }
    
    UIImage *image = [UIImage imageNamed:@"arrow"];
    if ([image isKindOfClass:UIImage.class]) {
        CGContextDrawImage(context, CGRectMake(size.width / 2 - 8, size.height / 2 - 8, 16, 16), image.CGImage);
    }

    UIImage *img = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    return img;
}

+ (UIImage *)decodedImageWithImage:(UIImage *)image {
    if (image.images) {
        // Do not decode animated images
        return image;
    }
    
    CGImageRef imageRef = image.CGImage;
    CGSize imageSize = CGSizeMake(CGImageGetWidth(imageRef), CGImageGetHeight(imageRef));
    CGRect imageRect = (CGRect){.origin = CGPointZero, .size = imageSize};
    
    CGColorSpaceRef colorSpace = CGColorSpaceCreateDeviceRGB();
    CGBitmapInfo bitmapInfo = CGImageGetBitmapInfo(imageRef);
    
    int infoMask = (bitmapInfo & kCGBitmapAlphaInfoMask);
    BOOL anyNonAlpha = (infoMask == kCGImageAlphaNone ||
                        infoMask == kCGImageAlphaNoneSkipFirst ||
                        infoMask == kCGImageAlphaNoneSkipLast);
    
    // CGBitmapContextCreate doesn't support kCGImageAlphaNone with RGB.
    // https://developer.apple.com/library/mac/#qa/qa1037/_index.html
    if (infoMask == kCGImageAlphaNone && CGColorSpaceGetNumberOfComponents(colorSpace) > 1) {
        // Unset the old alpha info.
        bitmapInfo &= ~kCGBitmapAlphaInfoMask;
        
        // Set noneSkipFirst.
        bitmapInfo |= kCGImageAlphaNoneSkipFirst;
    }
    // Some PNGs tell us they have alpha but only 3 components. Odd.
    else if (!anyNonAlpha && CGColorSpaceGetNumberOfComponents(colorSpace) == 3) {
        // Unset the old alpha info.
        bitmapInfo &= ~kCGBitmapAlphaInfoMask;
        bitmapInfo |= kCGImageAlphaPremultipliedFirst;
    }
    
    // It calculates the bytes-per-row based on the bitsPerComponent and width arguments.
    CGContextRef context = CGBitmapContextCreate(NULL,
                                                 imageSize.width,
                                                 imageSize.height,
                                                 CGImageGetBitsPerComponent(imageRef),
                                                 0,
                                                 colorSpace,
                                                 bitmapInfo);
    CGColorSpaceRelease(colorSpace);
    
    // If failed, return undecompressed image
    if (!context) return image;
    
    CGContextDrawImage(context, imageRect, imageRef);
    CGImageRef decompressedImageRef = CGBitmapContextCreateImage(context);
    
    CGContextRelease(context);
    
    UIImage *decompressedImage = [UIImage imageWithCGImage:decompressedImageRef scale:image.scale orientation:image.imageOrientation];
    CGImageRelease(decompressedImageRef);
    return decompressedImage;
}

+ (UIImage *)compressImage:(UIImage *)sourceImage targetWidth:(CGFloat)targetWidth {
    CGFloat sourceWidth = sourceImage.size.width;
    CGFloat sourceHeight = sourceImage.size.height;
    CGFloat targetHeight = (targetWidth / sourceWidth) * sourceHeight;
    
    CGFloat compressRate = sourceWidth * sourceHeight / (targetWidth * targetHeight);
    if (compressRate <= 1.0f) {
        return sourceImage;
    }
    
//    UIGraphicsBeginImageContext(CGSizeMake(targetWidth, targetHeight));
    UIGraphicsBeginImageContextWithOptions(CGSizeMake(targetWidth, targetHeight), NO, 1.0);
    [sourceImage drawInRect:CGRectMake(0,0,targetWidth, targetHeight)];
    UIImage *newImage = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    return newImage;
}

+ (UIImage *)compressImage:(UIImage *)image width:(NSInteger)minWidth {
    if (![image isKindOfClass:UIImage.class]) {
        return nil;
    }
    if (image.size.width < 1) {
        return image;
    }
    NSInteger comp = 1;
    NSInteger oriWith = image.size.width;
    NSInteger targetWidth = oriWith;
    while (targetWidth > minWidth) {
         comp *= 2;
        targetWidth = oriWith / comp;
    }
    if (targetWidth < minWidth) {
        comp /= 2;
    }
    if (comp < 2) {
        return image;
    }
    
    if (comp > 0) {
        NSInteger width = oriWith / comp;
        return [self compressImage:image targetWidth:width];
    }
    return image;
}

+ (UIImage *)imageForUIView:(nonnull UIView *)view {
    if (view == nil) {
        return nil;
    }
    
    CGSize size = view.bounds.size;
    CGRect savedFrame = view.frame;
    BOOL isScrollView = [view isKindOfClass:UIScrollView.class];
    if (isScrollView) {
        UIScrollView *sView = (UIScrollView *)view;
        size = CGSizeMake(sView.frame.size.width,sView.contentSize.height + sView.contentInset.top + sView.contentInset.bottom);
        view.frame = CGRectMake(view.frame.origin.x, view.frame.origin.y, size.width, size.height);
    }
    
    UIImage *image = nil;
    if (@available(iOS 10.0, *)) {
        UIGraphicsImageRendererFormat *format = [[UIGraphicsImageRendererFormat alloc] init];
        if (@available(iOS 12.0, *)) {
            format.preferredRange = UIGraphicsImageRendererFormatRangeStandard;
        }
        UIGraphicsImageRenderer *renderer = [[UIGraphicsImageRenderer alloc] initWithSize:view.bounds.size format:format];
        __weak typeof(view) weakView = view;
        image = [renderer imageWithActions:^(UIGraphicsImageRendererContext * _Nonnull rendererContext) {
            return [weakView.layer renderInContext:rendererContext.CGContext];
        }];
    }
    
    if (isScrollView) {
        view.frame = savedFrame;
    }
    return image;
}

+ (UIImage *)imageGradualFromColorWithARed:(CGFloat)aRed aGreen:(CGFloat)aGreen aBlue:(CGFloat)aBlue aAlpha:(CGFloat)aAlpha toColorWithBRed:(CGFloat)bRed bGreen:(CGFloat)bGreen bBlue:(CGFloat)bBlue bAlpha:(CGFloat)bAlpha width:(CGFloat)width height:(CGFloat)height {
    CGRect frame = CGRectMake(0, 0, width, height);
    UIGraphicsBeginImageContext(CGSizeMake(width, height));
    CGContextRef context = UIGraphicsGetCurrentContext();
    CGColorSpaceRef rgb = CGColorSpaceCreateDeviceRGB();
    CGContextScaleCTM(context, frame.size.width, frame.size.height);
    CGFloat colors[] = {
        aRed, aGreen, aBlue, aAlpha,
        bRed, bGreen, bBlue, bAlpha,
    };
    CGGradientRef backGradient = CGGradientCreateWithColorComponents(rgb, colors, NULL, sizeof(colors)/(sizeof(colors[0])*4));
    CGColorSpaceRelease(rgb);
    //设置颜色渐变的方向，范围在(0,0)与(1.0,1.0)之间，如(0,0)(1.0,0)代表水平方向渐变,(0,0)(0,1.0)代表竖直方向渐变
    CGContextDrawLinearGradient(context, backGradient, CGPointMake(0, 0), CGPointMake(1.0, 0), kCGGradientDrawsBeforeStartLocation);
    UIImage *image = UIGraphicsGetImageFromCurrentImageContext();
    return image;
}

@end

File: FSUIKit/Classes/Swift/FSImage.swift
// FSImageS.swift
// Translated from FSImage.h/m

import UIKit
import Photos

public class FSImage: NSObject {
    
    public static func image(from color: UIColor) -> UIImage? {
        return image(from: color, size: CGSize(width: 1, height: 10))
    }
    
    public static func image(from color: UIColor, size: CGSize) -> UIImage? {
        let rect = CGRect(origin: .zero, size: size)
        UIGraphicsBeginImageContext(rect.size)
        guard let context = UIGraphicsGetCurrentContext() else { return nil }
        context.setFillColor(color.cgColor)
        context.fill(rect)
        let img = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()
        return img
    }
    
    public static func image(withSize size: CGSize, direction isHorizon: Bool, offsetLeft left: CGFloat, offsetRight right: CGFloat, ratios: [NSNumber], colors: [UIColor]) -> UIImage? {
        guard !ratios.isEmpty, ratios.count == colors.count else { return nil }
        
        let rect = CGRect(origin: .zero, size: size)
        UIGraphicsBeginImageContext(rect.size)
        guard let context = UIGraphicsGetCurrentContext() else { return nil }
        
        var sum: CGFloat = 0
        for (index, ratio) in ratios.enumerated() {
            let fratio = CGFloat(ratio.floatValue)
            let color = colors[index]
            
            context.setFillColor(color.cgColor)
            if isHorizon {
                context.fill(CGRect(x: 0, y: sum * size.height, width: size.width, height: fratio * size.height))
            } else {
                context.fill(CGRect(x: sum * size.width, y: 0, width: fratio * size.width, height: size.height))
            }
            sum += fratio
        }
        
        let img = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()
        return img
    }
    
    public static func image(withSize size: CGSize, backgroundColor: UIColor?, mainColor: UIColor?, marginColor: UIColor?) -> UIImage? {
        let rect = CGRect(origin: .zero, size: size)
        UIGraphicsBeginImageContext(rect.size)
        guard let context = UIGraphicsGetCurrentContext() else { return nil }
        
        if let bgColor = backgroundColor {
            context.setFillColor(bgColor.cgColor)
            context.fill(CGRect(x: 0, y: 0, width: size.width * 0.25, height: size.height))
            context.fill(CGRect(x: size.width * 0.75, y: 0, width: size.width * 0.25, height: size.height))
        }
        
        if let main = mainColor {
            context.setFillColor(main.cgColor)
            context.fill(CGRect(x: size.width * 0.3125, y: 0, width: size.width * 0.375, height: size.height))
        }
        
        if let margin = marginColor {
            context.setFillColor(margin.cgColor)
            context.fill(CGRect(x: size.width * 0.25, y: 0, width: size.width * 0.0625, height: size.height))
            context.fill(CGRect(x: size.width * 0.6875, y: 0, width: size.width * 0.0625, height: size.height))
        }
        
        if let image = UIImage(named: "arrow") {
            context.draw(image.cgImage!, in: CGRect(x: size.width / 2 - 8, y: size.height / 2 - 8, width: 16, height: 16))
        }
        
        let img = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()
        return img
    }
    
    public static func decodedImage(with image: UIImage) -> UIImage {
        if image.images != nil { return image }
        
        guard let imageRef = image.cgImage else { return image }
        let imageSize = CGSize(width: CGFloat(imageRef.width), height: CGFloat(imageRef.height))
        let imageRect = CGRect(origin: .zero, size: imageSize)
        
        guard let colorSpace = CGColorSpace(name: CGColorSpace.sRGB) else { return image }
        var bitmapInfo = imageRef.bitmapInfo
        
        let infoMask = bitmapInfo.intersection(.alphaInfoMask)
        let anyNonAlpha = infoMask == CGBitmapInfo(rawValue: CGImageAlphaInfo.none.rawValue) ||
                          infoMask == CGBitmapInfo(rawValue: CGImageAlphaInfo.noneSkipFirst.rawValue) ||
                          infoMask == CGBitmapInfo(rawValue: CGImageAlphaInfo.noneSkipLast.rawValue)
        
        if infoMask == CGBitmapInfo(rawValue: CGImageAlphaInfo.none.rawValue) && colorSpace.numberOfComponents > 1 {
            bitmapInfo.remove(.alphaInfoMask)
            bitmapInfo.insert(CGBitmapInfo(rawValue: CGImageAlphaInfo.noneSkipFirst.rawValue))
        } else if !anyNonAlpha && colorSpace.numberOfComponents == 3 {
            bitmapInfo.remove(.alphaInfoMask)
            bitmapInfo.insert(CGBitmapInfo(rawValue: CGImageAlphaInfo.premultipliedFirst.rawValue))
        }
        
        guard let context = CGContext(data: nil, width: Int(imageSize.width), height: Int(imageSize.height), bitsPerComponent: imageRef.bitsPerComponent, bytesPerRow: 0, space: colorSpace, bitmapInfo: bitmapInfo.rawValue) else { return image }
        
        context.draw(imageRef, in: imageRect)
        guard let decompressedImageRef = context.makeImage() else { return image }
        
        return UIImage(cgImage: decompressedImageRef, scale: image.scale, orientation: image.imageOrientation)
    }
    
    public static func compressImage(_ sourceImage: UIImage, targetWidth: CGFloat) -> UIImage {
        let sourceWidth = sourceImage.size.width
        let sourceHeight = sourceImage.size.height
        let targetHeight = (targetWidth / sourceWidth) * sourceHeight
        
        let compressRate = sourceWidth * sourceHeight / (targetWidth * targetHeight)
        if compressRate <= 1.0 { return sourceImage }
        
        UIGraphicsBeginImageContextWithOptions(CGSize(width: targetWidth, height: targetHeight), false, 1.0)
        sourceImage.draw(in: CGRect(x: 0, y: 0, width: targetWidth, height: targetHeight))
        let newImage = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()
        return newImage ?? sourceImage
    }
    
    public static func compressImage(_ image: UIImage, width minWidth: Int) -> UIImage? {
        guard image.size.width >= 1 else { return image }
        
        var comp = 1
        let oriWidth = Int(image.size.width)
        var targetWidth = oriWidth
        
        while targetWidth > minWidth {
            comp *= 2
            targetWidth = oriWidth / comp
        }
        if targetWidth < minWidth {
            comp /= 2
        }
        if comp < 2 { return image }
        
        if comp > 0 {
            let width = oriWidth / comp
            return compressImage(image, targetWidth: CGFloat(width))
        }
        return image
    }
    
    public static func image(for view: UIView, useDrawHierarchy: Bool = true, afterScreenUpdates: Bool = true) -> UIImage? {
        let savedFrame = view.frame
        var savedContentOffset: CGPoint = .zero

        if let scrollView = view as? UIScrollView {
            savedContentOffset = scrollView.contentOffset

            let width = max(scrollView.bounds.width, scrollView.contentSize.width + scrollView.contentInset.left + scrollView.contentInset.right)
            let height = max(scrollView.bounds.height, scrollView.contentSize.height + scrollView.contentInset.top + scrollView.contentInset.bottom)
            view.frame = CGRect(x: savedFrame.origin.x, y: savedFrame.origin.y, width: width, height: height)
            scrollView.contentOffset = .zero
            scrollView.layoutIfNeeded()
        }

        defer {
            if let scrollView = view as? UIScrollView {
                scrollView.contentOffset = savedContentOffset
            }
            view.frame = savedFrame
        }

        let size = view.bounds.size
        guard size.width > 0, size.height > 0 else { return nil }

        let format = UIGraphicsImageRendererFormat.default()
        format.scale = UIScreen.main.scale
        format.opaque = false

        let renderer = UIGraphicsImageRenderer(size: size, format: format)
        let image = renderer.image { context in
            if useDrawHierarchy {
                let ok = view.drawHierarchy(in: view.bounds, afterScreenUpdates: afterScreenUpdates)
                if !ok {
                    view.layer.render(in: context.cgContext)
                }
            } else {
                view.layer.render(in: context.cgContext)
            }
        }

        return image
    }
    
    public static func imageGradual(fromColorWithARed aRed: CGFloat, aGreen: CGFloat, aBlue: CGFloat, aAlpha: CGFloat, toColorWithBRed bRed: CGFloat, bGreen: CGFloat, bBlue: CGFloat, bAlpha: CGFloat, width: CGFloat, height: CGFloat) -> UIImage? {
        UIGraphicsBeginImageContext(CGSize(width: width, height: height))
        guard let context = UIGraphicsGetCurrentContext() else { return nil }
        guard let rgb = CGColorSpace(name: CGColorSpace.sRGB) else { return nil }
        
        context.scaleBy(x: width, y: height)
        let colors: [CGFloat] = [aRed, aGreen, aBlue, aAlpha, bRed, bGreen, bBlue, bAlpha]
        guard let backGradient = CGGradient(colorSpace: rgb, colorComponents: colors, locations: nil, count: 2) else { return nil }
        
        context.drawLinearGradient(backGradient, start: CGPoint(x: 0, y: 0), end: CGPoint(x: 1.0, y: 0), options: .drawsBeforeStartLocation)
        let image = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()
        return image
    }

    public static func requestPhotoLibraryAddPermission(completion: @escaping (Bool) -> Void) {
        if #available(iOS 14, *) {
            let status = PHPhotoLibrary.authorizationStatus(for: .addOnly)
            if status == .authorized || status == .limited {
                completion(true)
                return
            }
            if status == .denied || status == .restricted {
                completion(false)
                return
            }
            PHPhotoLibrary.requestAuthorization(for: .addOnly) { newStatus in
                completion(newStatus == .authorized || newStatus == .limited)
            }
        } else {
            let status = PHPhotoLibrary.authorizationStatus()
            if status == .authorized {
                completion(true)
                return
            }
            if status == .denied || status == .restricted {
                completion(false)
                return
            }
            PHPhotoLibrary.requestAuthorization { newStatus in
                completion(newStatus == .authorized)
            }
        }
    }

    public static func saveToPhotoLibrary(_ image: UIImage, completion: @escaping (Bool) -> Void) {
        requestPhotoLibraryAddPermission { granted in
            guard granted else {
                completion(false)
                return
            }

            PHPhotoLibrary.shared().performChanges({
                PHAssetCreationRequest.creationRequestForAsset(from: image)
            }, completionHandler: { success, _ in
                completion(success)
            })
        }
    }
}
